File: app/src/main/java/com/weechan/asr/MainActivity.java
//package com.weechan.asr;
//
//import android.Manifest;
//import android.annotation.SuppressLint;
//import android.os.Bundle;
//import com.google.android.material.floatingactionbutton.FloatingActionButton;
//import androidx.core.app.ActivityCompat;
//import androidx.appcompat.app.AppCompatActivity;
//import androidx.recyclerview.widget.LinearLayoutManager;
//import androidx.recyclerview.widget.RecyclerView;
//import kotlin.Unit;
//import kotlin.jvm.functions.Function0;
//
//import android.view.Gravity;
//import android.view.MotionEvent;
//import android.view.View;
//import android.widget.PopupWindow;
//
//import com.weechan.asr.utils.permission.Permission;
//import com.weechan.asr.utils.permission.PermissionCompatActivity;
//import com.weechan.asr.utils.permission.PermissionMan;
//import com.weechan.asr.widget.WaveView;
//
//import java.io.File;
//
//import static android.view.MotionEvent.ACTION_UP;
//
//public class MainActivity extends PermissionCompatActivity {
//
//    private RecyclerView recyclerView;
//    private Adaptee adapter;
//
//    private FloatingActionButton fab;
//
//    SoundModel model = new SoundModel();
//    private PopupWindow po;
//    private WaveView popupWave;
//
//
//    @Override
//    protected void onCreate(Bundle savedInstanceState) {
//        super.onCreate(savedInstanceState);
//        setContentView(R.layout.activity_main);
//        Permission.STORAGE.doAfterGet(this, () -> {
//            Permission.AUDIO.doAfterGet(MainActivity.this,()->{
//
//            });
//            return null;
//        });
//        initView();
//    }
//
//    private void startRecord() {
//        po.showAtLocation(findViewById(R.id.container), Gravity.CENTER, 0, 0);
//        model.start(() -> popupWave.setWaves(model.getDatas(),true));
//    }
//
//    private void stopRecord() {
//        adapter.addWavesInActivePos(model.getDatas());
//        adapter.notifyDataSetChanged();
//        model.stop();
//        popupWave.setWaves(null,true);
//        po.dismiss();
//
//
//
//
////
////        new Thread(() -> {
////            long time = System.currentTimeMillis();
////            runOnUiThread(() -> {
////                Toast.makeText(this, "开始分 析,分析未完成请不要操作", Toast.LENGTH_SHORT).show();
////            });
////            File outFile = new File(curFile.getAbsolutePath() + ".wav");
////            AudioRecorder.convertPcmToWav(curFile.getAbsolutePath(), outFile.getAbsolutePath(), 16000, 1,   16);
////            Analyze.analyze(Environment.getExternalStorageDirectory().getPath() + "/SA2_.wav");
////            runOnUiThread(() -> {
////                Toast.makeText(this, "分析耗时　" + (System.currentTimeMillis() - time), Toast.LENGTH_SHORT).show();
////                StringBuilder result = new StringBuilder();
////                try {
////                    FileInputStream fis = new FileInputStream(new File(getFilesDir()+"/model", "output.txt"));
////                    BufferedReader br = new BufferedReader(new InputStreamReader(fis));
////                    String line;
////                    while ((line = br.readLine()) != null) {
////                        String[] two = line.split(",");
////                        int soundIndex = Integer.parseInt(two[0]);
////                        long pos = Long.parseLong(two[1]);
////                        String sound = soundMap.get(soundIndex);
////                        result.append(sound);
////                    }
////                } catch (FileNotFoundException e) {
////                    e.printStackTrace();
////                } catch (IOException e) {
////                    e.printStackTrace();
////                }
////                records.add(new Record(result.toString(), 1));
////                popupWave.setWaves(null, false);
////                adapter.notifyDataSetChanged();
////            });
////        }).start();
//    }
//
//    @SuppressLint("ClickableViewAccessibility")
//    private void initView() {
//
//        adapter = new Adaptee(model.records);
//        fab = findViewById(R.id.fab);
//        recyclerView = findViewById(R.id.recyclerView);
//        recyclerView.setLayoutManager(new LinearLayoutManager(this));
//        recyclerView.setAdapter(adapter);
//        recyclerView.setItemAnimator(null);
//
//        View WaveWrap = View.inflate(this, R.layout.popup_audio, null);
//        po = new PopupWindow(WaveWrap);
//        po.setWidth(dp2px(240));
//        po.setHeight(dp2px(120));
//        popupWave = WaveWrap.findViewById(R.id.wave);
//        popupWave.setMode(0);
//        popupWave.setWaves(model.getDatas(), true);
//
//        fab.setOnTouchListener((v, event) -> {
//            int action = event.getAction();
//            if (action == MotionEvent.ACTION_DOWN) {
//                startRecord();
//            }
//
//            if (action == ACTION_UP) {
//                stopRecord();
//            }
//
//            return true;
//        });
//
//    }
//
//    private int dp2px(final float dpValue) {
//        final float scale = getResources().getDisplayMetrics().density;
//        return (int) (dpValue * scale + 0.5f);
//    }
//}

File: app/src/main/java/com/weechan/asr/MainActivity.kt
package com.weechan.asr

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.widget.EditText
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.tencent.bugly.crashreport.CrashReport
import com.weechan.asr.base.MultiEntity
import com.weechan.asr.net.BASE_URL
import com.weechan.asr.utils.album.AlbumPickerActivity
import com.weechan.asr.utils.permission.Permission
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import org.jetbrains.anko.alert
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

class MainActivity : AlbumPickerActivity() {

    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: Adapter
    private lateinit var model: SoundModel


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        init()
        val pref = getSharedPreferences("ip", Context.MODE_PRIVATE)
        BASE_URL = "http://${pref.getString("ip","127.0.0.1")}"
        CrashReport.initCrashReport(applicationContext, "008ec33f99", true)
        button.setOnClickListener {
            alert {
                val et = EditText(this@MainActivity)
                et.setText(pref.getString("ip","127.0.0.1"))
                et.hint = pref.getString("ip","127.0.0.1")
                customView = et;
                title = "请输入IP"
                this.positiveButton("OK"){
                    BASE_URL = et.text.toString()
                    pref.edit().putString("ip",et.text.toString()).apply()
                }
            }.show()
        }
    }

    fun init() {
        GlobalScope.launch(Dispatchers.Main) {
            val r1 = suspendCoroutine<Boolean> {
                Permission.STORAGE.get(this@MainActivity) { r ->
                    it.resume(r)
                }
            }

            val r2 = suspendCoroutine<Boolean> {
                Permission.AUDIO.get(this@MainActivity) { r ->
                    it.resume(r)
                }
            }

            if (r1 && r2) {
                model = SoundModel()
                GlobalScope.async { model.init() }.await()
                initView()
            } else {
                init()
            }
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun initView() {
        adapter = Adapter(model.sources.map { MultiEntity(it,1) })
        recyclerView = findViewById(R.id.recyclerView)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = adapter
    }

}
